src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt

```diff
@@ -32,10 +32,18 @@ class AddMutableFix(binding: RsNamedElement) : LocalQuickFixAndIntentionActionOn
     companion object {
         fun createIfCompatible(expr: RsExpr): AddMutableFix? {
             val declaration = expr.declaration
-            return if (declaration is RsSelfParameter || (declaration is RsPatBinding && declaration.kind is BindByValue)) {
-                AddMutableFix(declaration as RsNamedElement)
-            } else {
-                null
+
+            return when {
+                declaration is RsSelfParameter -> {
+                    AddMutableFix(declaration as RsNamedElement)
+                }
+                expr.type is TyReference && declaration is RsPatBinding && declaration.isArg && !declaration.isMut -> {
```

Comment:
`isMut` is checked before the creation of the quick fix, so you shouldn't check it here. Also, I found the problem. Consider such code:
```Rust
fn foo(ref s: &S) {
    s.a = 42;
}
```
Converting the argument to `ref s: &mut S` doesn't help. rustc also doesn't suggest such fix to this code because it doesn't work.

 So you should check `declaration.kind is BindByValue` here, and then the resulted arms will be:
`declaration is RsPatBinding && declaration.kind is BindByValue && declaration.isArg`
`declaration is RsPatBinding && declaration.kind is BindByValue && expr.type !is TyReference`

Also, I'm not sure that these arms are completely correct for us, so it'd be really great if you test this fix on different cases of Rust code (and add such tests to `AddMutableFixTest`).

Reply:
I'll try to figure out something

Reply:
OK, you're welcome to create another PR if you find something is missed.